File: cpp/cppsearch/tests/SearchOptionsTest.cpp
#include <algorithm>
#include <catch2/catch_all.hpp>
#include <unordered_set>
#include "SearchOptions.h"

TEST_CASE("Get SearchSettings from minimal args", "[SearchOptions]") {
    auto* options = new cppsearch::SearchOptions();
    char* argv[] = { const_cast<char *>("cppsearch"), const_cast<char *>("-s"), const_cast<char *>("Searcher"),
                     const_cast<char *>(".") };
    int argc = 4;
    auto settings = options->settings_from_args(argc, argv);
    REQUIRE(!settings.archives_only());
    REQUIRE(!settings.debug());
    REQUIRE(!settings.first_match());
    REQUIRE(!settings.include_hidden());
    REQUIRE((settings.lines_after() == 0));
    REQUIRE((settings.lines_before() == 0));
    REQUIRE((settings.max_line_length() == 150));
    REQUIRE(!settings.multi_line_search());
    REQUIRE(!settings.print_dirs());
    REQUIRE(!settings.print_files());
    REQUIRE(!settings.print_lines());
    REQUIRE(settings.print_results());
    REQUIRE(!settings.print_usage());
    REQUIRE(!settings.print_version());
    REQUIRE(!settings.search_archives());
    REQUIRE(!settings.unique_lines());
    REQUIRE(!settings.verbose());

    REQUIRE(settings.in_archive_extensions().empty());
    REQUIRE(settings.in_archive_file_patterns().empty());
    REQUIRE(settings.in_dir_patterns().empty());
    REQUIRE(settings.in_extensions().empty());
    REQUIRE(settings.in_file_patterns().empty());
    REQUIRE(settings.out_archive_extensions().empty());
    REQUIRE(settings.out_archive_file_patterns().empty());
    REQUIRE(settings.out_dir_patterns().empty());
    REQUIRE(settings.out_extensions().empty());
    REQUIRE(settings.out_file_patterns().empty());

    auto search_patterns = settings.search_patterns();
    REQUIRE(search_patterns.size() == 1);
    std::unordered_set<std::string> expected_patterns{"Searcher"};
    auto contains_pattern = [&](auto& p) {
        return expected_patterns.contains(p.pattern());
    };
    REQUIRE(std::all_of(search_patterns.cbegin(), search_patterns.cend(), contains_pattern));

    REQUIRE(!settings.paths().empty());
    REQUIRE(settings.paths().size() == 1);
    REQUIRE(settings.paths().contains("."));
}

TEST_CASE("Get SearchSettings from valid args", "[SearchOptions]") {
    auto* options = new cppsearch::SearchOptions();
    char* argv[] = { const_cast<char *>("cppsearch"), const_cast<char *>("-x"), const_cast<char *>("java,scala"),
                     const_cast<char *>("-s"), const_cast<char *>("Searcher"), const_cast<char *>(".") };
    int argc = 6;
    auto settings = options->settings_from_args(argc, argv);
    REQUIRE(!settings.archives_only());
    REQUIRE(!settings.debug());
    REQUIRE(!settings.first_match());
    REQUIRE(settings.include_hidden());
    REQUIRE((settings.lines_after() == 0));
    REQUIRE((settings.lines_before() == 0));
    REQUIRE((settings.max_line_length() == 150));
    REQUIRE(!settings.multi_line_search());
    REQUIRE(!settings.print_dirs());
    REQUIRE(!settings.print_files());
    REQUIRE(!settings.print_lines());
    REQUIRE(settings.print_results());
    REQUIRE(!settings.print_usage());
    REQUIRE(!settings.print_version());
    REQUIRE(!settings.search_archives());
    REQUIRE(!settings.unique_lines());
    REQUIRE(!settings.verbose());

    REQUIRE(settings.in_archive_extensions().empty());
    REQUIRE(settings.in_archive_file_patterns().empty());
    REQUIRE(settings.in_dir_patterns().empty());
    REQUIRE(settings.in_file_patterns().empty());
    REQUIRE(settings.out_archive_extensions().empty());
    REQUIRE(settings.out_archive_file_patterns().empty());
    REQUIRE(settings.out_dir_patterns().empty());
    REQUIRE(settings.out_extensions().empty());
    REQUIRE(settings.out_file_patterns().empty());

    std::unordered_set<std::string> in_exts = settings.in_extensions();
    REQUIRE(in_exts.size() == 2);
    REQUIRE(in_exts.contains("java"));
    REQUIRE(in_exts.contains("scala"));

    auto search_patterns = settings.search_patterns();
    REQUIRE(search_patterns.size() == 1);
    std::unordered_set<std::string> expected_patterns{"Searcher"};
    auto contains_pattern = [&](auto& p) {
        return expected_patterns.contains(p.pattern());
    };
    REQUIRE(std::all_of(search_patterns.cbegin(), search_patterns.cend(), contains_pattern));

    REQUIRE(!settings.paths().empty());
    REQUIRE(settings.paths().size() == 1);
    REQUIRE(settings.paths().contains("."));
}

TEST_CASE("Get SearchSettings with archives-only", "[SearchOptions]") {
    auto* options = new cppsearch::SearchOptions();
    char* argv[] = { const_cast<char *>("cppsearch"), const_cast<char *>("-x"), const_cast<char *>("java,scala"),
                     const_cast<char *>("-s"), const_cast<char *>("Searcher"), const_cast<char *>("--archivesonly"),
                     const_cast<char *>(".") };
    int argc = 7;
    auto settings = options->settings_from_args(argc, argv);
    REQUIRE(settings.archives_only());
    REQUIRE(settings.search_archives());
}

TEST_CASE("Get SearchSettings from JSON", "[SearchOptions]") {
    std::string json = R"(
{
    "out-dirpattern": ["build", "node_module", "tests", "typings"],
    "out-filepattern": ["gulpfile", "\\.min\\."],
    "path": "~/src/xsearch/",
    "in-ext": ["js","ts"],
    "searchpattern": "Searcher",
    "linesbefore": 2,
    "linesafter": 2,
    "debug": true,
    "allmatches": false,
    "includehidden": true
}
)";

    auto options = cppsearch::SearchOptions();
    auto settings = cppsearch::SearchSettings();
    options.update_settings_from_json(settings, json);

    REQUIRE(!settings.paths().empty());
    REQUIRE(settings.paths().size() == 1);
    REQUIRE(settings.paths().contains("~/src/xsearch/"));
    REQUIRE(settings.in_extensions().size() == 2);
    REQUIRE(settings.in_extensions().contains("js"));
    REQUIRE(settings.in_extensions().contains("ts"));

    auto out_dir_patterns = settings.out_dir_patterns();
    REQUIRE(out_dir_patterns.size() == 4);
    std::unordered_set<std::string> expected_out_dir_patterns{"build", "node_module", "tests", "typings"};
    auto contains_out_dir_pattern = [&](auto& p) {
        return expected_out_dir_patterns.contains(p.pattern());
    };
    REQUIRE(std::all_of(out_dir_patterns.cbegin(), out_dir_patterns.cend(), contains_out_dir_pattern));

    auto out_file_patterns = settings.out_file_patterns();
    REQUIRE(out_file_patterns.size() == 2);
    std::unordered_set<std::string> expected_out_file_patterns{"gulpfile", "\\.min\\."};
    auto contains_out_file_pattern = [&](auto& p) {
        return expected_out_file_patterns.contains(p.pattern());
    };
    REQUIRE(std::all_of(out_file_patterns.cbegin(), out_file_patterns.cend(), contains_out_file_pattern));

    auto search_patterns = settings.search_patterns();
    REQUIRE(search_patterns.size() == 1);
    std::unordered_set<std::string> expected_search_patterns{"Searcher"};
    auto contains_pattern = [&](auto& p) {
        return expected_search_patterns.contains(p.pattern());
    };
    REQUIRE(std::all_of(search_patterns.cbegin(), search_patterns.cend(), contains_pattern));

    REQUIRE(settings.lines_before() == 2);
    REQUIRE(settings.lines_after() == 2);
    REQUIRE(settings.debug() == true);
    REQUIRE(settings.first_match() == true);
    REQUIRE(settings.include_hidden() == true);
}

File: dart/dartsearch/test/search_options_test.dart
import 'package:dartsearch/dartsearch.dart' show SearchOptions, SearchSettings;
import 'package:test/test.dart';

void main() {
  test('test get settings from minimal args', () async {
    var options = SearchOptions();
    var settings = await options.settingsFromArgs(['-s', 'Search', '.']);
    expect(settings.archivesOnly, false);
    expect(settings.colorize, true);
    expect(settings.debug, false);
    expect(settings.firstMatch, false);
    expect(settings.includeHidden, true);
    expect(settings.linesAfter, 0);
    expect(settings.linesBefore, 0);
    expect(settings.maxLineLength, 150);
    expect(settings.multiLineSearch, false);
    expect(settings.printDirs, false);
    expect(settings.printFiles, false);
    expect(settings.printLines, false);
    expect(settings.printResults, true);
    expect(settings.printUsage, false);
    expect(settings.printVersion, false);
    expect(settings.searchArchives, false);
    expect(settings.recursive, true);
    expect(settings.uniqueLines, false);
    expect(settings.verbose, false);
    expect(settings.searchPatterns.length, 1);
    expect((settings.searchPatterns.first as RegExp).pattern, 'Search');
    expect(settings.paths.length, 1);
    expect(settings.paths.first, '.');
  });

  test('test get settings from valid args', () async {
    var options = SearchOptions();
    var settings =
        await options.settingsFromArgs(['-x', 'dart,kt', '-s', 'Search', '.']);
    expect(settings.inExtensions.length, 2);
    expect(settings.inExtensions.contains('dart'), true);
    expect(settings.inExtensions.contains('kt'), true);
    expect(settings.searchPatterns.length, 1);
    expect((settings.searchPatterns.first as RegExp).pattern, 'Search');
    expect(settings.paths.length, 1);
    expect(settings.paths.first, '.');
  });

  test('test get settings with archives-only', () async {
    var options = SearchOptions();
    var args = ['-x', 'dart,kt', '-s', 'Search', '--archivesonly', '.'];
    var settings = await options.settingsFromArgs(args);
    expect(settings.archivesOnly, true);
    expect(settings.searchArchives, true);
  });

  test('test get settings from json', () async {
    var json = '{'
        '"path": "~/src/xsearch/",'
        '"in-ext": ["js","ts"],'
        '"out-dirpattern": ["build", "node_module", "tests", "typings"],'
        r'"out-filepattern": ["gulpfile", "\\.min\\."],'
        '"searchpattern": "Searcher",'
        '"linesbefore": 2,'
        '"linesafter": 2,'
        '"debug": true,'
        '"allmatches": false,'
        '"includehidden": true'
        '}';
    var options = SearchOptions();
    var settings = SearchSettings();
    await options.settingsFromJson(json, settings);

    expect(settings.paths.first, '~/src/xsearch/');

    expect(settings.inExtensions.length, 2);
    expect(settings.inExtensions.contains('js'), true);
    expect(settings.inExtensions.contains('ts'), true);

    expect(settings.outDirPatterns.length, 4);
    expect(
        settings.outDirPatterns
            .any((p) => (p as RegExp).pattern == 'node_module'),
        true);

    expect(settings.outFilePatterns.length, 2);
    expect(
        settings.outFilePatterns
            .any((p) => (p as RegExp).pattern == 'gulpfile'),
        true);

    expect(settings.searchPatterns.length, 1);
    expect((settings.searchPatterns.first as RegExp).pattern, 'Searcher');

    expect(settings.linesBefore, 2);
    expect(settings.linesAfter, 2);

    expect(settings.debug, true);
    expect(settings.verbose, true);
    expect(settings.firstMatch, true);
    expect(settings.includeHidden, true);
  });
}
